drivers/display/ek79652.c

```diff
@@ -0,0 +1,552 @@
+/*
+ * Copyright (c) 2021 Linumiz
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gooddisplay_ek79652
+
+#include <string.h>
+#include <device.h>
+#include <init.h>
+#include <drivers/display.h>
+#include <drivers/gpio.h>
+#include <drivers/spi.h>
+#include <sys/byteorder.h>
+
+#include "ek79652_regs.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(ek79652, CONFIG_DISPLAY_LOG_LEVEL);
+
+/**
+ * EK79652 compatible EPD controller driver.
+ *
+ */
+#define EK79652_DC_PIN			DT_INST_GPIO_PIN(0, dc_gpios)
+#define EK79652_DC_FLAGS		DT_INST_GPIO_FLAGS(0, dc_gpios)
+#define EK79652_DC_CNTRL		DT_INST_GPIO_LABEL(0, dc_gpios)
+#define EK79652_BUSY_PIN		DT_INST_GPIO_PIN(0, busy_gpios)
+#define EK79652_BUSY_CNTRL		DT_INST_GPIO_LABEL(0, busy_gpios)
+#define EK79652_BUSY_FLAGS		DT_INST_GPIO_FLAGS(0, busy_gpios)
+#define EK79652_RESET_PIN		DT_INST_GPIO_PIN(0, reset_gpios)
+#define EK79652_RESET_CNTRL		DT_INST_GPIO_LABEL(0, reset_gpios)
+#define EK79652_RESET_FLAGS		DT_INST_GPIO_FLAGS(0, reset_gpios)
+
+#define EPD_PANEL_WIDTH			DT_INST_PROP(0, width)
+#define EPD_PANEL_HEIGHT		DT_INST_PROP(0, height)
+#define EK79652_PIXELS_PER_BYTE		8U
+
+/* Horizontally aligned page! */
+#define EK79652_NUMOF_PAGES		(EPD_PANEL_WIDTH / \
+					 EK79652_PIXELS_PER_BYTE)
+#define EK79652_BUFFER_SIZE		(EK79652_NUMOF_PAGES * EPD_PANEL_HEIGHT)
+#define EK79652_PANEL_FIRST_GATE	0U
+#define EK79652_PANEL_LAST_GATE		(EPD_PANEL_HEIGHT - 1)
+#define EK79652_PANEL_FIRST_PAGE	0U
+#define EK79652_PANEL_LAST_PAGE		(EK79652_NUMOF_PAGES - 1)
+
+struct ek79652_data {
+	const struct ek79652_config *config;
+	const struct device *reset;
+	const struct device *dc;
+	const struct device *busy;
+	bool initial;
+	bool full_refresh;
+	bool disp_refresh;
+};
+
+struct ek79652_config {
+	struct spi_dt_spec bus;
+};
+
+static uint8_t ek79652_softstart[] = DT_INST_PROP(0, softstart);
+static uint8_t ek79652_pwr[] = DT_INST_PROP(0, pwr);
+#if DT_INST_NODE_HAS_PROP(0, pwropt)
+static uint8_t ek79652_pwropt[] = DT_INST_PROP(0, pwropt);
+#endif
+
+static int ek79652_write_data(struct ek79652_data *driver,
+				   uint8_t cmd, uint8_t *data, size_t len,
+				   uint8_t *disp_data, size_t disp_len)
+{
+	struct spi_buf buf = {.buf = &cmd, .len = sizeof(cmd)};
+	struct spi_buf_set buf_set = {.buffers = &buf, .count = 1};
+
+	gpio_pin_set(driver->dc, EK79652_DC_PIN, 1);
+	if (spi_write_dt(&driver->config->bus, &buf_set)) {
+		return -EIO;
+	}
+
+	if (data != NULL) {
+		for (int i = 0 ; i < len ; i++) {
+			buf.buf = &data[i];
+			buf.len = 1;
+			gpio_pin_set(driver->dc, EK79652_DC_PIN, 0);
+			if (spi_write_dt(&driver->config->bus, &buf_set)) {
+				return -EIO;
+			}
+		}
+	}
+
+	if (disp_data != NULL) {
```

Comment:
Why so cumbersome? Why not just _ek79652_write_cmd()_, as in gd7965 driver from which you have taken over so many lines.

Reply:
Because according to [datasheet](https://ibb.co/FDBjLKQ)  inbetween sending bytes CS pin should be made HIGH.  Since CS pin need to be made LOW only when sending data, this display is working only when sending byte by byte data. In zephyr we can't toggle CS pin. so only _ek79652_write_data()_ function is introduced.